src/hotspot/share/gc/g1/g1ConcurrentMark.cpp

```diff
@@ -1176,111 +1177,155 @@ void G1ConcurrentMark::verify_during_pause(G1HeapVerifier::G1VerifyType type,
   }
 }
 
-class G1UpdateRemSetTrackingBeforeRebuildTask : public WorkerTask {
+// Update per-region liveness info based on CM stats. Then, reclaim empty
+// regions right away and select certain regions (e.g. sparse ones) for remset
+// rebuild.
+class G1UpdateRegionLivenessAndSelectForRebuildTask : public WorkerTask {
   G1CollectedHeap* _g1h;
   G1ConcurrentMark* _cm;
   HeapRegionClaimer _hrclaimer;
+
   uint volatile _total_selected_for_rebuild;
 
-  G1PrintRegionLivenessInfoClosure _cl;
+  // Reclaimed empty regions
+  FreeRegionList _cleanup_list;
 
-  class G1UpdateRemSetTrackingBeforeRebuild : public HeapRegionClosure {
+  struct G1OnRegionClosure : public HeapRegionClosure {
     G1CollectedHeap* _g1h;
     G1ConcurrentMark* _cm;
+    // The number of regions actually selected for rebuild.
+    uint _num_selected_for_rebuild;
 
-    G1PrintRegionLivenessInfoClosure* _cl;
+    size_t _freed_bytes;
+    uint _num_old_regions_removed;
+    uint _num_humongous_regions_removed;
+    FreeRegionList* _local_cleanup_list;
 
-    uint _num_regions_selected_for_rebuild;  // The number of regions actually selected for rebuild.
+    G1OnRegionClosure(G1CollectedHeap* g1h,
+                      G1ConcurrentMark* cm,
+                      FreeRegionList* local_cleanup_list) :
+      _g1h(g1h),
+      _cm(cm),
+      _num_selected_for_rebuild(0),
+      _freed_bytes(0),
+      _num_old_regions_removed(0),
+      _num_humongous_regions_removed(0),
+      _local_cleanup_list(local_cleanup_list) {}
 
-    void update_remset_before_rebuild(HeapRegion* hr) {
-      G1RemSetTrackingPolicy* tracking_policy = _g1h->policy()->remset_tracker();
+    void reclaim_empty_humongous_region(HeapRegion* hr) {
+      assert(!hr->has_pinned_objects(), "precondition");
+      assert(hr->is_starts_humongous(), "precondition");
 
-      bool selected_for_rebuild;
-      if (hr->is_humongous()) {
-        bool const is_live = _cm->contains_live_object(hr->humongous_start_region()->hrm_index());
-        selected_for_rebuild = tracking_policy->update_humongous_before_rebuild(hr, is_live);
-      } else {
-        size_t const live_bytes = _cm->live_bytes(hr->hrm_index());
-        selected_for_rebuild = tracking_policy->update_before_rebuild(hr, live_bytes);
-      }
-      if (selected_for_rebuild) {
-        _num_regions_selected_for_rebuild++;
-      }
-      _cm->update_top_at_rebuild_start(hr);
+      auto on_humongous_region = [&] (HeapRegion* hr) {
+        assert(hr->used() > 0, "precondition");
+        _num_humongous_regions_removed++;
+        _freed_bytes += hr->used();
+        hr->set_containing_set(nullptr);
```

Comment:
```suggestion
        assert(hr->used() > 0, "precondition");
        assert(!hr->has_pinned_objects(), "precondition");
        assert(hr->is_humongous(), "precondition");

        _num_humongous_regions_removed++;

        _freed_bytes += hr->used();
        hr->set_containing_set(nullptr);
```
Just a nit to make the `reclaim_empty*` methods look more equal. The pinned check can now be added because it means something different than before when humongous regions were implicitly pinned.